base/strings/io.jl

```diff
@@ -2,8 +2,25 @@
 
 ## core text I/O ##
 
-print(io::IO, x) = show(io, x)
-print(io::IO, xs...) = for x in xs print(io, x) end
+function print(io::IO, x)
+    lock(io)
+    try
+        show(io, x)
+    finally
+        unlock(io)
+    end
+end
+
+function print(io::IO, xs...)
+    lock(io)
+    try
+        for x in xs
+            print(io, x)
+        end
+    finally
+        unlock(io)
+    end
+end
```

Comment:
> Yeah but it used to apply only to AsyncStreams; now it will be used everywhere.

that's the kind of incorrect logic that got us into the broken mess that this PR attempts to fix. there was never a clear difference between AsyncStream and IO, other than some places randomly only accepted AsyncStream and some places only worked if given an AsyncStream -- and the two were not consistent or uniform.